### PR TITLE
Implementation of rule 3.6 - line break style

### LINKED ISSUE #120: Rule 3.6 line break style rules if the line is split
1. Compared to Java Kotlin allows not to put semicolon (';') after each statement separated by newline.
 There should be no redundant semicolon at the end of lines.
 
 In case when newline is needed to split the line, it should be placed after operators like &&/||/+/e.t.c
 But newline should be placed before operators like ('.', '?.', '?:', '::', e.t.c).
 
* Note, that you need to follow functional style (i.e. each function call in a chain with `.` should start at a new line if chain of functions contains more than one call):
```kotlin
  val value = otherValue!!
          .map { x -> x }
          .filter {
              val a = true
              true
          }
          .size    
```
* Note, that Kotlin parser prohibits operator !! to be separated from the value it is checking.

**Exception**: if functional chain is used inside of the branches of ternary operator - no need to split them with newlines. Valid example:
```kotlin
if (condition) list.map { foo(it) }.filter { bar(it) } else list.drop(1)
```
   
2. Newline should be placed after assignment operator ('=')
3. The name of a function or constructor should not be split by a newline from the opening brace '('. Brace follows immediately after the name without any spaces.   
4. Newline should be placed right after the comma (',')
5. In lambda statements, if it's body contains more than one line, the newline should be placed after an arrow. See examples below:

Bad example:
```kotlin
    value.map { name -> foo()
        bar()
    }
```

Recommended examples：
```kotlin
value.map { name ->
    foo()
    bar()
}

val someValue = { node:String -> node }
```

6. When the function contains only a single expression, it can be expressed as [expression function] (https://kotlinlang.org/docs/reference/functions.html#single-expression-functions).
Instead of: 
```kotlin
override fun toString(): String { return "hi" }
```
use:
```kotlin
override fun toString() = "hi"
```

### PR DESCRIPTION
**Pull request by petertrr**
### What's done:
* Implemented NewlinesRule
* Added tests
* Extracted several functions from SingleLineStatementsRule to AstNodeUtils
* Minor refactoring
* **Some edits** in code style guide

## Which rule and warnings did you add?

This pull request closes #120

## Actions checklist
* [x] Implemented Rule, added Warnings
* [x] Added tests on checks
* [x] Added tests on fixers
* [x] Updated rules-config.json
* [x] Updated available-rules.md

## Fixme

* [x] Finish ignored tests
* [x] Implement exception case for functional style

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+ * 2. Checks that some operators are followed by newline, while others are prepended by it
+ * 3. Statements that follow `!!` behave in the same way
+ * 4. Forces functional style of chained dot call expressions with exception
+ */
+@Suppress("ForbiddenComment")
+class NewlinesRule : Rule("newlines") {
+    companion object {
+        // fixme: these token sets can be not full, need to add new once as corresponding cases are discovered.
+        // error is raised if these operators are prepended by newline
+        private val lineBreakAfterOperators = TokenSet.create(ANDAND, OROR, PLUS, PLUSEQ, MINUS, MINUSEQ, MUL, MULTEQ, DIV, DIVEQ)
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:60]: xor?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        } else if (node.elementType in lineBreakBeforeOperators && !node.isDotFromPackageOrImport()) {
+            handleOperatorWithLineBreakBefore(node)
+        }
+    }
+
+    private fun handleSemicolon(node: ASTNode) {
+        if (node.isEol() && node.treeParent.elementType != ENUM_ENTRY) {
+            // semicolon at the end of line which is not part of enum members declarations
+            REDUNDANT_SEMICOLON.warnAndFix(configRules, emitWarn, isFixMode, node.extractLineOfText(), node.startOffset) {
+                node.treeParent.removeChild(node)
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:94]: treeParent always exists?

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/NewlinesRuleWarnTest.kt [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                LintError(3, 26, ruleId, "$functionalStyleWarn .", true),
+                LintError(5, 10, ruleId, "$functionalStyleWarn ?.", true)
+        )
+    }
+
+    @Test
+    fun `chained calls should follow functional style - exception for ternary if-else`() {
+        lintMethod(NewlinesRule(),
+                """
+                    |fun foo(list: List<Bar>?) {
```
Comment (akuleshov7) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/NewlinesRuleWarnTest.kt:173]: lets add regression test for the future with the function and argument list?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt (hunk 2 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:94]: Even if file consists of single semicolon, it's `treeParent` will be file itself, so should be safe.

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/NewlinesRuleWarnTest.kt (hunk 3 above)
Comment (petertrr) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/NewlinesRuleWarnTest.kt:173]: Yes, will add tests and mark them `@Disabled`.

**Pull request merged**